# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

- **Consider a significance level of 5% for all tests.**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as st

# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [2]:
pokemon = pd.read_csv('../pokemon.csv')

In [3]:
df = pokemon.copy()

Let's start off by looking at the `head` function in the cell below.

In [4]:
df.head(3)

#       Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1  Bulbasaur  Grass  Poison    318  45      49       49       65       65   
1  2    Ivysaur  Grass  Poison    405  60      62       63       80       80   
2  3   Venusaur  Grass  Poison    525  80      82       83      100      100   

   Speed  Generation  Legendary  
0     45           1      False  
1     60           1      False  
2     80           1      False

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [5]:
legendary = df[df["Legendary"]== True]["Total"]
non_legendary = df[df["Legendary"]== False]["Total"]

In [6]:
print(f'Legendary pokemons are: {legendary.value_counts().sum()}')
print(f'Non legendary pokemons: {non_legendary.value_counts().sum()}')

Legendary pokemons are: 65
Non legendary pokemons: 735


Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [7]:
df_legend_vs_non=pd.DataFrame({"legendary":legendary,"non legendary": non_legendary})

df_legend_vs_non.describe().head(3)

legendary  non legendary
count   65.000000     735.000000
mean   637.384615     417.213605
std     60.937389     106.760417

The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

H0: legendary = non legendary



In [8]:
alpha = 0.05

In [9]:
t_statistic,p_value = st.ttest_ind(legendary,non_legendary, equal_var= False)
t_statistic,p_value

(25.8335743895517, 9.357954335957446e-47)

What do you conclude from this test? Write your conclusions below.

In [10]:
if p_value < alpha:
    print("Reject null hipothesis. Legendary pokemons are significantly diferents to no legendary")
else:
    print("Non reject null hipothesis. Legendary pokemons aren´t significantly diferents to no legendary")


Reject null hipothesis. Legendary pokemons are significantly diferents to no legendary


How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [11]:
unique_types = df["Type 1"].unique()
type_totals = {}

In [12]:
for type_name in unique_types:
    type_data = df[df["Type 1"] == type_name]
    type_totals[type_name] = type_data["Total"].values

In [13]:
type_totals = df.groupby("Type 1")["Total"].sum()

In [14]:
tytot = type_totals.sort_values(ascending = False)

In [15]:
dftype = pd.DataFrame(tytot)
dftype.T

Type 1  Water  Normal  Grass  Psychic    Bug   Fire   Rock  Electric  Dragon  \
Total   48211   39365  29480    27129  26146  23820  19965     19510   17617   

Type 1  Ghost  Ground   Dark  Steel  Fighting  Poison    Ice  Fairy  Flying  
Total   14066   14000  13818  13168     11244   11176  10403   7024    1940

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [16]:
water = df[df["Type 1"] == "Water"]["Total"]
others = df[df["Type 1"] != "Water"]["Total"]

mean_w = water.mean()
std_w = water.std()

mean_o = others.mean()
std_o = others.std()

In [17]:
print(f"The mean of water pokemons is {mean_w}")
print(f"The standar deviation of water pokemons is {std_w}")
print(f"The mean of other types of pokemons is {mean_o}")
print(f"The mean of other types of pokemons is {std_o}")

The mean of water pokemons is 430.45535714285717
The standar deviation of water pokemons is 113.18826606431458
The mean of other types of pokemons is 435.85901162790697
The mean of other types of pokemons is 121.09168230208066


The mean represents the average "Total" attribute value for each type of Pokémon in their respective groups. In this context, it indicates that, on average, Pokémon of type "Water" have a slightly lower "Total" value compared to Pokémon of other types.

The standard deviation measures the spread of values around the mean. In this context, it suggests that the "Total" attribute values for Pokémon of type "Water" tend to have less variation compared to Pokémon of other types, as the standard deviation is lower.

Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

H0: mean_w = mean_other (Means aren´t differents estatistically, it means taking account the std.)

H1: mean_w != mean_other (Means are differents estatistically, it means taking account the std.)

In [18]:
t_statistic, p_value = st.ttest_ind(water,others, equal_var=True)
t_statistic, p_value

(-0.4418547448849676, 0.6587140317488793)

In [19]:
if  p_value < 0.05:
    print("Reject null hipothesis")
else:
    print("We cannot reject null hipothesis.")
          

We cannot reject null hipothesis.


Write your conclusion below.

Mean of water pokemons are statistally the same than others, with a small mean varianz of -0.4418547448849676.



# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [20]:
attack = df["Attack"]
defense = df["Defense"]

H0: Attack = Defense (statistically)

H1: Attack != Defense (statistically)

In [21]:
t_statistic, p_value = st.ttest_rel(attack,defense)
t_statistic, p_value

(4.325566393330478, 1.7140303479358558e-05)

In [22]:
if  p_value < 0.05:
    print("Reject null hipothesis")
else:
    print("We cannot reject null hipothesis")

Reject null hipothesis


Describe the results of the test in the cell below.

Since the p-value (0.0000171) is much smaller than the typical significance level of 0.05, there is sufficient evidence to reject the null hypothesis that the means of "attack" and "defense" are equal.
Therefore, we can conclude that there is a significant difference between the mean values of "attack" and "defense" in the studied population.
The t-statistic value is approximately 4.33.
This value indicates the magnitude of the difference between the means of the two paired sample



We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [23]:
spattack = df["Sp. Atk"]
spdefense = df["Sp. Def"]

H0: Sp. Atk = Sp. Def (statistically)

H1: Sp. Atk != Sp. Def (statistically)

In [24]:
t_statistic, p_value = st.ttest_rel(spattack,spdefense)
t_statistic, p_value

(0.853986188453353, 0.3933685997548122)

In [25]:
if  p_value < 0.05:
    print("Reject null hipothesis")
else:
    print("We cannot reject null hipothesis")

We cannot reject null hipothesis


Describe the results of the test in the cell below.

This value is the probability of observing such an extreme difference between the means of the paired samples if the null hypothesis is true (i.e., if the samples come from populations with equal means)



As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [26]:
AD_ratio = df["Attack"] - df["Defense"]

H0: Attack mean - Def mean = 0

H1: Attack mean - Def mean != 0

In [27]:
t_statistic, p_value = st.ttest_1samp(AD_ratio, 0)
t_statistic, p_value

(4.325566393330478, 1.7140303479358558e-05)

In [28]:
if  p_value < 0.05:
    print("Reject null hipothesis")
else:
    print("We cannot reject null hipothesis")

Reject null hipothesis


# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [29]:
# Your code here:



Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [30]:
# Your code here:



Based on a 95% confidence, should we reject the null hypothesis?

In [31]:
# Your answer here:

